### ISSUE TITLE
Crash using lua sleep examples

### ISSUE THREAD

**Issue by c-taylor** [apache/trafficserver#7372]
ATS version 8.1.x

I was able to consistently cause server restarts when using the ts.sleep function as per examples in the admin guide.

### Configuration
```
map http://property1.mycdn.com/    http://origin1.mycdn.com:8080/  @plugin=tslua.so @pparam=sleep-request.lua
map https://property1.mycdn.com/    http://origin1.mycdn.com:8080/  @plugin=tslua.so @pparam=sleep-request.lua
```

sleep-request.lua:
```
    function send_response()
        ts.sleep(3)
    end
    function read_response()
        ts.sleep(3)
    end
    function do_remap()
        ts.hook(TS_LUA_HOOK_READ_RESPONSE_HDR, read_response)
        ts.hook(TS_LUA_HOOK_SEND_RESPONSE_HDR, send_response)
    end
```

Hooks that crash:
“TS_LUA_HOOK_(SEND|READ)_RESPONSE_HDR”
“TS_LUA_HOOK _CACHE_LOOKUP_COMPLETE”
“TS_LUA_HOOK_SEND_REQUEST_HDR”

Hooks that don't crash:
“TS_LUA_RESPONSE_TRANSFORM”
“TS_LUA_HOOK_TXN_CLOSE”
“TS_LUA_HOOK _OS_DNS”

Sleep only really performed as expected for "TS_LUA_HOOK _OS_DNS” in my case.
There is also an. example in 'plugins/lua/example/test_sleep.lua', derivations of which have the same results.

**Comment by shukitchan** [comment]
Can you share the stack trace when it crashed ?
Also I understand that i understand that it crashed only when some work loads are put onto it. Can you elaborate on that, too? Thanks.

**Comment by c-taylor** [comment]
I’ll check on the stack trace...

Workload was quite simple: The first few requests in a cache hit/miss spread <1 second I would estimate.

**Comment by c-taylor** [comment]
Stack trace attached: A few redactions, let me know if it's too mangled.

[sleep_lua_stack.txt](https://github.com/apache/trafficserver/files/5655649/sleep_lua_stack.txt)

**Comment by shukitchan** [comment]
Mark told me that it happened with over 1000+ RPS on that machine . So you are saying that a few requests only will cause the crash? 

I think I can reproduce this . But I have a simple setup on my mac/minikube/ATS on docker. I can do 100+ RPS with no problem and get a similar crash as you a few seconds later if i put 200+ RPS load on the setup.

**Comment by c-taylor** [comment]
Sorry, I never tried to identify if this was a request rate, concurrency or total with respect to triggering.

It was sub-1 second into a load test that was 500 req/sec. It sounds like you have a trigger, so I won't add more for now.

**Comment by shukitchan** [comment]
It looks like it is related to thread affinity for the continuation. Let me try to use the same thread for the new continuation for sleep and see.

**Comment by shukitchan** [comment]
Apparently it looks like i can't set thread affinity to the calling thread in 8.1.x 
The API and mechanism are available in 9.0.x, though. Will try that out later

**Comment by shukitchan** [comment]
Actually just try it out in 9.0.0 . It works fine. 
I think it is due to the change mentioned here - https://docs.trafficserver.apache.org/en/latest/developer-guide/api/functions/TSContScheduleOnPool.en.html?highlight=tscontscheduleonpool

ts.sleep() is implemented using TSContScheduleOnPool(). The idea is to schedule a continuation to wake up again later according to the specified duration and let that re-enable the one that runs ts.sleep() . And if the thread is different, it will cause the crash. And the change in TSContScheduleOnPool in 9.0.0 is such that the thread will be the same (i.e. scenario 2 - no thread affinity info, same types of threads)
